llvm/lib/SYCLLowerIR/LowerESIMDVLoadVStore.cpp

```diff
@@ -0,0 +1,142 @@
+//===- LowerESIMDVLoadVStore.cpp - lower vload/vstore to load/store -------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// convert vload/vstore to load/store if they are not for genx_volatile.
+//
+// File scope simd variables marked with genx_volatile attribute want
+// guarranteed allocation in register file, therefore we use vload/vstore
+// instead of load/store, so they won't be optimized away by llvm.
+//
+// For ordinary simd variables, we do not need to protect load/store. But
+// there is no good way to do this in clang. So we need this pass in the
+// end of module passes to separate the cases that we need vload/vstore vs.
+// the cases that we do not need vload/vstore
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "loweresimdvloadvstore"
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/GenXIntrinsics/GenXIntrinsics.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/IR/Module.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/Transforms/Scalar.h"
+
+#include "llvm/Pass.h"
+
+using namespace llvm;
+
+namespace llvm {
+void initializeESIMDLowerLoadStorePass(PassRegistry &);
+}
+
+namespace {
+
+struct ESIMDLowerLoadStore : public FunctionPass {
+  static char ID;
+  ESIMDLowerLoadStore() : FunctionPass(ID) {
+    initializeESIMDLowerLoadStorePass(*PassRegistry::getPassRegistry());
+  }
+  virtual void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+  }
+
+  virtual bool runOnFunction(Function &F) override;
+
+private:
+  bool lowerLoadStore(Function &F);
+};
+
+} // namespace
+
+char ESIMDLowerLoadStore::ID = 0;
+INITIALIZE_PASS(ESIMDLowerLoadStore, "ESIMDLowerLoadStore",
+                "Lower ESIMD reference loads and stores", false, false)
+
+bool ESIMDLowerLoadStore::runOnFunction(Function &F) {
+  bool Changed = false;
+  Changed |= lowerLoadStore(F);
+  return Changed;
+}
+
+// Lower non-volatilE vload/vstore intrinsic calls into normal load/store
+// instructions.
+bool ESIMDLowerLoadStore::lowerLoadStore(Function &F) {
+  // lower all vload/vstore into normal load/store.
+  std::vector<Instruction *> ToErase;
+  for (Instruction &Inst : instructions(F)) {
+    if (GenXIntrinsic::isVLoadStore(&Inst)) {
```

Comment:
This loop body is one huge block which is hard to read.

Could you please use more early exits there? I think it will be much easier to read. For example:
```
if (!GenXIntrinsic::isVLoadStore(&Inst))
      continue;
```
and so on.

Reply:
I think this is just your personal preference. Other people may prefer not to use continue

Reply:
This is not my personal preference. Here is the link to the LLVM coding guidelines:
https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

Reply:
Coding style is just a guideline. Individual developer should have some freedom in applying it. In this case, for example, I do not like the negative sense in the condition when using condition. I prefer to use positive sense whenever possible for the condition test. Clearly, you read this code, and it did not confuse you.

Reply:
Individual reviewers should have freedom in insisting to apply such guidelines (which are essentially a set of BKMs to write robust/correct/maintainable code) if they seem them suitable. I'm exercising that freedom.

Reply:
I will resolve this so we can move on.

As someone new to this project and this process, I do have some different thoughts on the code review process. Let me quote a piece I found  on the internet to support my concern:
"There’s a imbalanced power dynamic in code review between reviewer and developer. This imbalance of power is by design – one of the functions of code review is to serve as a pedagogical tool – bringing developers up-to-speed on a project’s standards quickly. Power imbalances in code review, if unchecked and persistent:
•	Demotivate existing contributors
•	Repel new contributions
•	Have a chilling effect on contributions in review
These outcomes all result in:
•	Fewer new contributors
•	Fewer contributions from existing contributors
•	A lower diversity of contributors
•	An overall slower pace of change "

So, Is there a way to avoid the hazards of power imbalance?

Reply:
Ok, I understand, but I am not the person who determines the review process for this project.
@pvchupin @olegmaslovatintel @bader @romanovvlad Could you please help with this question?

Reply:
This project follows LLVM coding guidelines and this is requirement.

Reply:
I have used "continue"

Reply:
Probably you forgot to upload the new changes.

Reply:
Not sure what I did wrong. It should be up to date now:  https://github.com/kbobrovs/llvm/blob/esimd-ldst/llvm/lib/SYCLLowerIR/LowerESIMDVLoadVStore.cpp